File: test/cudatest.py
import devito, devitopro
import numpy as np 
from devitopro.types.enriched import NoLayers

compression = 'bitcomp'
dtype=np.float32
nt = 10
grid = devito.Grid(shape=(50, 50), dtype=dtype)

u = devitopro.TimeFunction(name='u', grid=grid)
v = devitopro.TimeFunction(name='v', grid=grid)
usave = devitopro.TimeFunction(name='usave', grid=grid, save=nt, space='local',
                        compression=compression, layers=NoLayers)
vsave = devitopro.TimeFunction(name='vsave', grid=grid, save=nt, layers=NoLayers)

# Forward operator (compress)
eqns0 = [
    devito.Eq(u.forward, u + 1.),  # PDE
    devito.Eq(usave, u)            # u.dt2*v
]
op0 = devito.Operator(eqns0, opt=('advanced'))

# Backward operator (decompress)
eqns1 = [
    devito.Eq(v, usave),    # Adjoint PDE
    devito.Eq(vsave, v),    # xcor
]
op1 = devito.Operator(eqns1, opt=('advanced'))

op0.apply(time_M=nt)
op1.apply(time_M=nt-1)

File: test/cudatest.jl
using Devito

isdir("AAA") && rm("AAA", recursive=true, force=true)
dumpdir = mkdir("AAA")

compression = "bitcomp"

nt = 10
g = Grid(shape=(9, 9))
t = time_dim(g)

usave = TimeFunction(name="usave", grid=g, save=nt,
                        compression="bitcomp", serialization=dumpdir)

# Forward, dump everything to disk inside `dumpdir`
eq = Eq(usave, t)
op0 = Operator(eq)
apply(op0, time_M=nt-1)


# Now backward -- will fetch the data from disk
# We gonna use a different TimeFunction, to emulate the use case in which
# the overarching application, for whatever reason, can't reuse the
# same `usave`
v = TimeFunction(name="v", grid=g)
vsave = TimeFunction(name="vsave", grid=g, save=nt)

fetchdir = Devito.str2serial(Devito.serial2str(usave))

usave1 = TimeFunction(name="usave2", grid=g, save=nt,
                        compression="bitcomp", serialization=fetchdir)

eqns = [Eq(v, backward(v) - 1),
        Eq(vsave, usave1)]
op1 = Operator(eqns)
apply(op1, time_M=nt-1, time_m=0)